hadoop-hdds/docs/content/design/decommissioning.md

```diff
@@ -0,0 +1,720 @@
+  maintenance        | 0
+  in-flight          | 2
+  replicaCount       | 1
+
+We have on emissing replicas, but and node A can not be DECOMMISSIONED as we have no HEALTHY replicas at all.
+
+
+## Statefulness
+
+SCM stores all the node state in-memory. After a restart on the SCM side the datanode state can be lost.
+
+**Ozone doesn't guarantee that decommissioning/maintenance mode state survives the SCM restarts!!!**
+
+ * If SCM restarts DECOMMISSIONED nodes will not report any more container reports and the nodes won't be registered.
+ * ENTERING_MAINTENANCE and DECOMMISSIONING nodes will became HEALTHY again and the decommissioning CLI command should be repeated.
+ *  IN_MAINTENANCE nodes will become DEAD and all the containers will be replicated.
+
+ *Ozone assumes that the maintenance mode is used short-term and SCM is not restarted during this specific period.*
+
+
+*Reasoning*:
+
+Neither of the node state nor the container state are persisted in SCM side. The decommissioned state can be stored on the SCM side (or on the SCM side and the datanode side) which can provide better user experience (and may be implemented).
+
+But to support maintenance mode after restart all the container information is required to be persisted (which is a too big architectural change).
+
+To make a replication decision replication manager needs the number of healthy replicas (they are reported via heartbeats) AND the number of containers on the node which is in maintenance mode. The later one is not available if the SCM is restarted as the container map exists only in the memory and the node which is turned off can't report any more container reports. Therefore the information about the existing containers on the node which is in the maintenance mode **can't be available'**. 
+
+## Throttling
+
+SCM should avoid to request too many replication to live enough network bandwidth for the requests.
+
+Replication Manager can easily throttle the replication requests based on `inflightReplication` map.
+
+The first algorithm would specify the maximum number of the requests per node which should be checked before sendint out the replication request to the datanode.
+
+With the same approach more advanced calculation can be easily implemented (cluster-wide limit, topology-related limits, etc.)
+
+## User interface
+
+The decommissioning and maintenance mode can be administered with a CLI interface.
+
+Required feature:
+
+ * Set the state of a datanode (to DECOMMISSIONING or ENTERING_MAINTENANCE)
+ * Undo the decommissioning process 
+ * check the current progress: 
+   * This can be a table with the nodes, status of the nodes, number of containers, containers under replication and containers which doesn't much the stop condition yet (required replications)
+ * All the commands can support topology related filters (eg. display the nodes only for a specific rack or show the status of the nodes of s specific rack)
+  
+## Checks before the decommissioning
+
```

Comment:
This is probably 'nice to have' rather than essential. Perhaps a version 2 requirement?

Reply:
Depends from the complexity. If it's easy, we can add it. I think we can create subtasks in jira and define the first priorities. For example proper support for showing progress also can be missing from the very first version. This document can include the whole vision...

Reply:
There is a range of complexities for this. Eg:

5 node cluster -> Decom 3 nodes -> Not possible as remaining_nodes < 3.

That is very simple, but then you may have a cluster with 2 racks, decom all hosts on 1 rack - this may violate rack placement policy.

Is there enough space on the remaining nodes to house all the new replicas?

Can maintenance mode be entered immediately for a set of nodes, or will replication be needed which will take some time?

etc.

Perhaps we should have some short circuited check logic from day 1 that always returns OK to get the framework in place, and then add rules over time etc.

I guess we will never be able to cover all scenarios and the picture can change as soon as decom starts if some nodes go dead unexpectedly etc, so it will never be perfect.

Reply:
There two design invariant we have used to design this feature. That is;  
1. A Client I/O must never fail in case of maintenance mode.
2. Decommission is complete if and only if the healthy count equals the expected count for a container or a block.

With the framework of these rules:
> 5 node cluster -> Decom 3 nodes -> Not possible as remaining_nodes < 3.
Well, we should not do that since this violates rule 2. We cannot guarantee that Healthy replica count will equal to expected count since we allow only one replica of a container per machine.
At this point, Decommissioning with the safe parameters of the cluster is not possible. We should fail the decommissioning process rather than proceed. User can always switch off the node and be in the same position. I submit that we should not do something that is not safe, and give the impression to the user that it can be done.

>That is very simple, but then you may have a cluster with 2 racks, decom all hosts on 1 rack - this may violate rack placement policy.
Again, another case where the structural integrity of the cluster is violated. Let us think in terms of a Database; say you created a rule that a field is unique, and then you tried to insert value 1 twice. What is the expected behavior for the system? I submit that the system should fail the operation and let the user know that this operation violates the structural integrity of the system.

>Is there enough space on the remaining nodes to house all the new replicas?
Yes, another classic case where I would expect the cluster to do the right thing. So yes, this should be something that I should be told and the decomm operation failed if you cannot make enough copies.

>Perhaps we should have some short circuited check logic from day 1 that always returns OK to get the framework in place, and then add rules over time etc.
Under Ozone we have a history of doing this; the safe mode manager is a repository of rules; you can add rules and they are evaluated before exiting the system. We can do the very same; yes, we might not nail down all the rules, and in future, more rules will be added. 

Generally, when we design features for Ozone, we follow a simple rule:
1. Correctness
2. Performance
3. Usage.
If we come across a case where one of them is in conflict with the other, this gives us the right framework to choose the behavior for Ozone.

So in the case, there are a class of cases (and some complex and unknown) which are known and relatively easy for us handle; not handling them violates the correctness constraint.

>I guess we will never be able to cover all scenarios and the picture can change as soon as decom starts if some nodes go dead unexpectedly etc, so it will never be perfect.

Completely agree. The problem with any dynamic system. It is possible that initial checks will pass and during replication, we might fail to find a home for a container since a data node has failed; but that is something that replica manager will let us know. We will write the initial checks very similar to SafeModeManager -- and then proceed from there.

Now the question, should we write this first or last; I personally think that defining these rules and perhaps even writing them down is very use and will influence the design of the APIs that need in the SCM. So I perfer that this is solved earlier than later; but I am open to alternative solutions like doing it little later. Though I predict something like this will never get done "later" and will be addressed only when a real data loss gets reported from a customer.